File: Solutions/Python/Problem 010.py
import math

def question():
    print("""The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

Find the sum of all the primes below two million.""")

def sieve_of_eratosthenes(num):
    """
    Sieve of Eratosthenes: Find all prime numbers up to the given limit.
    Time Complexity: O(n * log(log(n))), where 'n' is the upper limit or the range of numbers up to which you want to find prime numbers.

    Args:
        num (int): The upper limit for finding prime numbers.

    Returns:
        int: The sum of all prime numbers below the limit.
    """
    base = 2
    num_list = list(range(2, num))
    while base * base < num:
        num_list = [i for i in num_list if i % base != 0 or i == base]
        base += 1
    return sum(num_list)

def sieve_of_sundaram(num):
    """
    Sieve of Sundaram: Find all prime numbers up to the given limit.
    Time Complexity: O(n * log(log(n))), where 'n' is the upper limit or the range of numbers up to which you want to find prime numbers.

    Args:
        num (int): The upper limit for finding prime numbers.

    Returns:
        int: The sum of all prime numbers below the limit.
    """
    bound = (num - 2) // 2
    print(bound)

def sieve_of_atkins(num):
    """
    Sieve of Atkins: Find all prime numbers up to the given limit.
    Time Complexity: O(n / log(log(n))), where 'n' is the upper limit

    Args:
        num (int): The upper limit for finding prime numbers.

    Returns:
        int: The sum of all prime numbers below the limit.
    """
    return True

def main():
    question()
    bound = 2000000
    print(f"The answer is {sieve_of_eratosthenes(bound)}")
    # Print("Explanation about the 3 sieves")

if __name__ == "__main__":
    main()

File: Solutions/Rust/problem_010.rs
// Problem 10 - Summation of primes
// https://projecteuler.net/problem=10
// Answer = 142913828922

fn question() {
    println!("Find the sum of all the primes below two million.");
}

fn solve(limit: usize) -> u64 {
    let mut is_prime = vec![true; limit];
    is_prime[0] = false;
    is_prime[1] = false;

    let mut i: usize = 2;
    while i * i < limit {
        if is_prime[i] {
            for j in (i * i..limit).step_by(i) {
                is_prime[j] = false;
            }
        }
        i += 1;
    }

    is_prime.into_iter().enumerate().filter_map(|(index, is_prime)| is_prime.then_some(index as u64)).sum()
}

fn main() {
    question();
    let limit = 2_000_000;
    println!("Answer = {}", solve(limit));
}
